parachain/src/wasm_api.rs

```diff
@@ -0,0 +1,63 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utilities for writing parachain WASM.
+
+use codec::{Encode, Decode};
+use super::{ValidationParams, ValidationResult, Message};
+
+mod ll {
+	extern "C" {
+		pub(super) fn ext_post_message(target: u32, data_ptr: *const u8, data_len: u32);
+	}
+}
+
+/// Load the validation params from memory when implementing a Rust parachain.
+///
+/// Offset and length must have been provided by the validation
+/// function's entry point.
```

Comment:
Since creation of a slice with null pointer is invalid maybe we should also mention it in this docs.

Reply:
oh, it is actually is required to be aligned as well. So because getting zero `offset` here is unlikely I'd just panic here.

Reply:
Don't the docs here cover all of that implicitly as long as we are providing good data to the validation function? maybe a bit overbroadly, but there is really only one valid use-case of this function.

Reply:
Yeah, I think I'd agree here!